actinia_module_plugin/core/modules/grass.py

```diff
@@ -79,39 +82,111 @@ def createModuleList(self):
+def build_and_run_iface_description_pc(self, module_list):
     pc = {"version": 1,
           "list": []}
 
     process_chain_items = []
     count = 1
     for module in module_list:
+        if type(module) is str:
+            module_id = module
+        else:
+            module_id = module['id']
         process_chain_items.append(
-            {"id": str(count), "module": module['id'],
+            {"id": str(count), "module": module_id,
              "interface-description": True})
         count = count + 1
 
     pc['list'] = process_chain_items
+
+    if len(module_list) == 1 and module in ["exporter", "importer"]:
+        # old style process chain was used before by createGrassModule
+        # as for importer and exporter new style fails, overwrite pc
+        # see https://github.com/mundialis/actinia_core/issues/227
+        pc = {"1": {"module": module, "interface-description": True}}
+
     response = run_process_chain(self, pc)
 
+    return response['process_log']
+
+
+def connect():
+    """This method initializes the connection with redis.
+    """
+    conf = Configuration()
+    try:
+        conf.read()
+    except Exception:
+        pass
+
+    server = conf.REDIS_SERVER_URL
+    port = conf.REDIS_SERVER_PORT
+    if conf.REDIS_SERVER_PW:
+        redis_password = conf.REDIS_SERVER_PW
+    else:
+        redis_password = None
+
+    redis_grass_module_interface.connect(
+        host=server, port=port, password=redis_password)
+
+    return redis_grass_module_interface
+
+
+def cacheGrassModule(grass_module):
+    '''
+    Insert grass_module into database
+    '''
+    redis_grass_module_interface = connect()
+    cached_module = redis_grass_module_interface.create(grass_module)
+
+    return cached_module
+
+
+def createGrassModule(self, module):
+    module_list = [module]
+    process_log = build_and_run_iface_description_pc(self, module_list)
+
+    xml_string = process_log[0]['stdout']
+    grass_module = ParseInterfaceDescription(xml_string)
+
+    # Check if cached module exists, then update if it has changed
+    # This is currently the only mechanism to update a cached
+    # module via API call.
+    # If not, then cache it
+    redis_grass_module_interface = connect()
+    if redis_grass_module_interface.exists(module):
+        redis_grass_module_interface.update(module, grass_module)
+    else:
+        cacheGrassModule(grass_module)
+
+    return grass_module
+
+
+def createFullModuleList(self, module_list):
     detailed_module_list = []
-    for desc in response['process_log']:
-        try:
-            grass_module = ParseInterfaceDescription(desc['stdout'])
+    request_module_list = []
+
+    # Check if modules are already cached, then use them
+    for module in module_list:
+        redis_grass_module_interface = connect()
+        grass_module = redis_grass_module_interface.read(module['id'])
+        if grass_module:
             detailed_module_list.append(grass_module)
```

Comment:
With large data sets like rasters the `append` can lead to runtime problems. If this list can also become very long, it might help if you create the list empty before and then fill it without having to change the length of the list.

Reply:
Thx for the hint. The list can be as long as the number of GRASS GIS modules. As it contains only python objects and this code runs in a synchronous HTTP request and would run into a timeout if it takes too long, I would leave it as is and if errors occur, I will remember.